File: javascript/tests/a2a.test.js
/**
 * Tests for v1.4 alpha.3 canonicalization id + A2A verification context.
 */

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

import { KeyManager, SignatureManager } from '../src/crypto.js';
import { SchemaPinCore } from '../src/core.js';
import {
    ErrorCode,
    KeyPinStore,
    verifySchemaOffline,
    verifySchemaForA2A,
    checkCanonicalization,
    CANONICALIZATION_V1,
    A2A_MAX_DELEGATION_DEPTH
} from '../src/verification.js';
import {
    A2aVerificationContext,
    a2aAllows,
    a2aIntersect,
    a2aIsUnrestricted
} from '../src/a2a.js';

// ─────────────────────────────────────────────────────────────────────
// AllowedDomains helpers (mirrors AgentPin v0.3 §4.11 semantics)
// ─────────────────────────────────────────────────────────────────────

describe('AllowedDomains helpers', () => {
    it('empty is unrestricted', () => {
        assert.equal(a2aIsUnrestricted([]), true);
        assert.equal(a2aIsUnrestricted(null), true);
        assert.equal(a2aIsUnrestricted(undefined), true);
    });

    it('unrestricted allows anything', () => {
        assert.equal(a2aAllows([], 'literally-anything'), true);
    });

    it('restricted filters', () => {
        const ad = ['api.client.com', '*.partner.com'];
        assert.equal(a2aAllows(ad, 'api.client.com'), true);
        assert.equal(a2aAllows(ad, 'tools.partner.com'), true);
        assert.equal(a2aAllows(ad, 'partner.com'), false); // *.partner.com excludes bare
        assert.equal(a2aAllows(ad, 'evil.example.com'), false);
    });

    it('intersect with unrestricted returns other', () => {
        assert.deepEqual(a2aIntersect([], ['a.com', 'b.com']), ['a.com', 'b.com']);
        assert.deepEqual(a2aIntersect(['a.com', 'b.com'], []), ['a.com', 'b.com']);
    });

    it('intersect returns overlap', () => {
        assert.deepEqual(
            a2aIntersect(['a.com', 'b.com', 'c.com'], ['b.com', 'c.com', 'd.com']),
            ['b.com', 'c.com']
        );
    });

    it('intersect empty overlap is unrestricted per spec', () => {
        const result = a2aIntersect(['a.com'], ['b.com']);
        assert.deepEqual(result, []);
        assert.equal(a2aIsUnrestricted(result), true);
    });
});

// ─────────────────────────────────────────────────────────────────────
// Canonicalization algorithm identifier
// ─────────────────────────────────────────────────────────────────────

describe('checkCanonicalization', () => {
    it('absent is supported', () => {
        assert.equal(checkCanonicalization(null), null);
        assert.equal(checkCanonicalization(undefined), null);
    });

    it('v1 is supported', () => {
        assert.equal(checkCanonicalization(CANONICALIZATION_V1), null);
        assert.equal(checkCanonicalization('schemapin-v1'), null);
    });

    it('unknown returns offending value', () => {
        assert.equal(checkCanonicalization('schemapin-v999'), 'schemapin-v999');
    });
});

// ─────────────────────────────────────────────────────────────────────
// Signed-schema fixture
// ─────────────────────────────────────────────────────────────────────

function setupSchema() {
    const { privateKey, publicKey } = KeyManager.generateKeypair();
    const pubPem = KeyManager.exportPublicKeyPem(publicKey);
    const schema = {
        name: 'calculate_sum',
        description: 'Calculates the sum of two numbers',
        parameters: { a: 'integer', b: 'integer' }
    };
    const schemaHash = SchemaPinCore.canonicalizeAndHash(schema);
    const sig = SignatureManager.signSchemaHash(schemaHash, privateKey);
    return {
        schema,
        sig,
        discovery: {
            schema_version: '1.2',
            developer_name: 'Test Developer',
            public_key_pem: pubPem,
            revoked_keys: []
        }
    };
}

// ─────────────────────────────────────────────────────────────────────
// verifySchemaOffline with canonicalization parameter
// ─────────────────────────────────────────────────────────────────────

describe('verifySchemaOffline canonicalization', () => {
    it('absent canonicalization accepted', () => {
        const f = setupSchema();
        const result = verifySchemaOffline(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore()
        );
        assert.equal(result.valid, true);
    });

    it('v1 canonicalization accepted', () => {
        const f = setupSchema();
        const result = verifySchemaOffline(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            CANONICALIZATION_V1
        );
        assert.equal(result.valid, true);
    });

    it('unknown canonicalization rejected', () => {
        const f = setupSchema();
        const result = verifySchemaOffline(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            'schemapin-v999'
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.CANONICALIZATION_UNSUPPORTED);
    });
});

// ─────────────────────────────────────────────────────────────────────
// verifySchemaForA2A
// ─────────────────────────────────────────────────────────────────────

function ctx(trusted, depth = 0) {
    return new A2aVerificationContext({
        callerAgentId: 'urn:agentpin:caller.com:test',
        delegationDepth: depth,
        originatingDomain: 'caller.com',
        trustedDomains: trusted
    });
}

describe('verifySchemaForA2A', () => {
    it('unrestricted caller allows any provider', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx([])
        );
        assert.equal(result.valid, true, JSON.stringify(result));
    });

    it('caller allow-list includes provider', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx(['example.com', 'other.com'])
        );
        assert.equal(result.valid, true);
    });

    it('provider outside caller scope rejected', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx(['other.com'])
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.A2A_SCOPE_VIOLATION);
    });

    it('delegation_depth cap enforced', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx([], A2A_MAX_DELEGATION_DEPTH + 1)
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.A2A_SCOPE_VIOLATION);
    });

    it('underlying signature failure passes through', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, 'bm90LWEtdmFsaWQtc2lnbmF0dXJl', 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx([])
        );
        assert.equal(result.valid, false);
        // Underlying error surfaces, not A2A_SCOPE_VIOLATION
        assert.equal(result.error_code, ErrorCode.SIGNATURE_INVALID);
    });

    it('canonicalization unknown rejected through A2A', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, f.sig, 'example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx([]),
            'schemapin-v999'
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.CANONICALIZATION_UNSUPPORTED);
    });

    it('wildcard provider in caller trusted list', () => {
        const f = setupSchema();
        const result = verifySchemaForA2A(
            f.schema, f.sig, 'api.example.com', 'calculate_sum',
            f.discovery, null, new KeyPinStore(),
            ctx(['*.example.com'])
        );
        assert.equal(result.valid, true, JSON.stringify(result));
    });
});

File: go/pkg/verification/a2a_test.go
package verification

import (
	"testing"

	"github.com/ThirdKeyAi/schemapin/go/pkg/core"
	"github.com/ThirdKeyAi/schemapin/go/pkg/crypto"
	"github.com/ThirdKeyAi/schemapin/go/pkg/discovery"
)

// ──────────────────────────────────────────────────────────────────────
// AllowedDomains helpers — AgentPin v0.3 §4.11 semantics
// ──────────────────────────────────────────────────────────────────────

func TestA2AIsUnrestricted(t *testing.T) {
	if !A2AIsUnrestricted(nil) {
		t.Error("nil should be unrestricted")
	}
	if !A2AIsUnrestricted([]string{}) {
		t.Error("empty slice should be unrestricted")
	}
	if A2AIsUnrestricted([]string{"a.com"}) {
		t.Error("non-empty slice should not be unrestricted")
	}
}

func TestA2AAllowsUnrestricted(t *testing.T) {
	if !A2AAllows(nil, "literally-anything") {
		t.Error("nil list should allow anything")
	}
}

func TestA2AAllowsRestricted(t *testing.T) {
	ad := []string{"api.client.com", "*.partner.com"}
	cases := map[string]bool{
		"api.client.com":         true,
		"tools.partner.com":      true,
		"deep.tools.partner.com": true,
		"partner.com":            false, // *.partner.com excludes bare
		"evil.example.com":       false,
	}
	for domain, want := range cases {
		got := A2AAllows(ad, domain)
		if got != want {
			t.Errorf("A2AAllows(%v, %q) = %v, want %v", ad, domain, got, want)
		}
	}
}

func TestA2AIntersectWithUnrestrictedReturnsOther(t *testing.T) {
	restricted := []string{"a.com", "b.com"}
	got := A2AIntersect(nil, restricted)
	if len(got) != 2 || got[0] != "a.com" || got[1] != "b.com" {
		t.Errorf("unrestricted ∩ X should equal X, got %v", got)
	}
	got = A2AIntersect(restricted, nil)
	if len(got) != 2 || got[0] != "a.com" || got[1] != "b.com" {
		t.Errorf("X ∩ unrestricted should equal X, got %v", got)
	}
}

func TestA2AIntersectOverlap(t *testing.T) {
	got := A2AIntersect([]string{"a.com", "b.com", "c.com"}, []string{"b.com", "c.com", "d.com"})
	want := []string{"b.com", "c.com"}
	if len(got) != len(want) || got[0] != want[0] || got[1] != want[1] {
		t.Errorf("expected %v, got %v", want, got)
	}
}

func TestA2AIntersectEmptyOverlapIsUnrestricted(t *testing.T) {
	got := A2AIntersect([]string{"a.com"}, []string{"b.com"})
	if !A2AIsUnrestricted(got) {
		t.Errorf("disjoint non-empty intersection should be re-interpreted as unrestricted, got %v", got)
	}
}

// ──────────────────────────────────────────────────────────────────────
// Canonicalization id
// ──────────────────────────────────────────────────────────────────────

func TestCheckCanonicalizationAbsent(t *testing.T) {
	if got := CheckCanonicalization(""); got != "" {
		t.Errorf("empty should be supported, got %q", got)
	}
}

func TestCheckCanonicalizationV1(t *testing.T) {
	if got := CheckCanonicalization(CanonicalizationV1); got != "" {
		t.Errorf("v1 should be supported, got %q", got)
	}
}

func TestCheckCanonicalizationUnknown(t *testing.T) {
	if got := CheckCanonicalization("schemapin-v999"); got != "schemapin-v999" {
		t.Errorf("unknown should return offending value, got %q", got)
	}
}

// ──────────────────────────────────────────────────────────────────────
// VerifySchemaForA2A — signed fixture helper
// ──────────────────────────────────────────────────────────────────────

type signedFixture struct {
	schema map[string]interface{}
	sig    string
	disc   *discovery.WellKnownResponse
}

func setupSignedSchema(t *testing.T) *signedFixture {
	t.Helper()
	kp, err := crypto.NewKeyManager().GenerateKeypair()
	if err != nil {
		t.Fatal(err)
	}
	pubPEM, err := crypto.NewKeyManager().ExportPublicKeyPEM(&kp.PublicKey)
	if err != nil {
		t.Fatal(err)
	}
	schema := map[string]interface{}{
		"name":        "calculate_sum",
		"description": "Calculates the sum of two numbers",
		"parameters":  map[string]interface{}{"a": "integer", "b": "integer"},
	}
	c := core.NewSchemaPinCore()
	schemaHash, err := c.CanonicalizeAndHash(schema)
	if err != nil {
		t.Fatal(err)
	}
	sig, err := crypto.NewSignatureManager().SignHash(schemaHash, kp)
	if err != nil {
		t.Fatal(err)
	}
	return &signedFixture{
		schema: schema,
		sig:    sig,
		disc: &discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "Test Developer",
			PublicKeyPEM:  pubPEM,
		},
	}
}

func ctxFor(trusted []string, depth uint8) *A2AVerificationContext {
	return &A2AVerificationContext{
		CallerAgentID:     "urn:agentpin:caller.com:test",
		DelegationDepth:   depth,
		OriginatingDomain: "caller.com",
		TrustedDomains:    trusted,
	}
}

// ──────────────────────────────────────────────────────────────────────
// verifySchemaOfflineWithCanonicalization
// ──────────────────────────────────────────────────────────────────────

func TestVerifySchemaOfflineCanonicalizationAbsent(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaOfflineWithCanonicalization(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(), "",
	)
	if !result.Valid {
		t.Fatalf("expected valid, got %+v", result)
	}
}

func TestVerifySchemaOfflineCanonicalizationV1(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaOfflineWithCanonicalization(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(), CanonicalizationV1,
	)
	if !result.Valid {
		t.Fatalf("expected valid, got %+v", result)
	}
}

func TestVerifySchemaOfflineCanonicalizationUnknown(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaOfflineWithCanonicalization(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(), "schemapin-v999",
	)
	if result.Valid {
		t.Fatal("expected invalid")
	}
	if result.ErrorCode != ErrCanonicalizationUnsupported {
		t.Errorf("expected %s, got %s", ErrCanonicalizationUnsupported, result.ErrorCode)
	}
}

// ──────────────────────────────────────────────────────────────────────
// VerifySchemaForA2A
// ──────────────────────────────────────────────────────────────────────

func TestVerifySchemaForA2AUnrestrictedCaller(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor(nil, 0), "",
	)
	if !result.Valid {
		t.Fatalf("expected valid, got %+v", result)
	}
}

func TestVerifySchemaForA2ACallerAllowsProvider(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor([]string{"example.com", "other.com"}, 1), "",
	)
	if !result.Valid {
		t.Fatalf("expected valid, got %+v", result)
	}
}

func TestVerifySchemaForA2AProviderOutsideScope(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor([]string{"other.com"}, 0), "",
	)
	if result.Valid {
		t.Fatal("expected invalid")
	}
	if result.ErrorCode != ErrA2AScopeViolation {
		t.Errorf("expected %s, got %s", ErrA2AScopeViolation, result.ErrorCode)
	}
}

func TestVerifySchemaForA2ADelegationDepthCap(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor(nil, A2AMaxDelegationDepth+1), "",
	)
	if result.Valid {
		t.Fatal("expected invalid")
	}
	if result.ErrorCode != ErrA2AScopeViolation {
		t.Errorf("expected %s, got %s", ErrA2AScopeViolation, result.ErrorCode)
	}
}

func TestVerifySchemaForA2AUnderlyingFailurePassesThrough(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, "bm90LWEtdmFsaWQtc2lnbmF0dXJl", "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor(nil, 0), "",
	)
	if result.Valid {
		t.Fatal("expected invalid")
	}
	if result.ErrorCode != ErrSignatureInvalid {
		t.Errorf("expected %s (underlying failure), got %s", ErrSignatureInvalid, result.ErrorCode)
	}
}

func TestVerifySchemaForA2ACanonicalizationUnknownPropagates(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, f.sig, "example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor(nil, 0), "schemapin-v999",
	)
	if result.Valid {
		t.Fatal("expected invalid")
	}
	if result.ErrorCode != ErrCanonicalizationUnsupported {
		t.Errorf("expected %s, got %s", ErrCanonicalizationUnsupported, result.ErrorCode)
	}
}

func TestVerifySchemaForA2AWildcardCallerAllowList(t *testing.T) {
	f := setupSignedSchema(t)
	result := VerifySchemaForA2A(
		f.schema, f.sig, "api.example.com", "calculate_sum",
		f.disc, nil, NewKeyPinStore(),
		ctxFor([]string{"*.example.com"}, 0), "",
	)
	if !result.Valid {
		t.Fatalf("expected valid, got %+v", result)
	}
}
